nav2_shim_controller/src/shim_controller.cpp

```diff
@@ -0,0 +1,397 @@
+  angle_to_path = atan2(carrot_pose.pose.position.y, carrot_pose.pose.position.x);
+  return fabs(angle_to_path) > angle_thresh;
+}
+
+
+bool ShimController::shouldRotateToGoalHeading(
+  const geometry_msgs::msg::PoseStamped & carrot_pose)
+{
+  // Whether we should rotate robot to goal heading
+  double dist_to_goal = std::hypot(carrot_pose.pose.position.x, carrot_pose.pose.position.y);
+  return dist_to_goal < goal_dist_tol_;
+}
+
+void ShimController::rotateToHeading(
+  double & angular_vel,
+  const double & angle_to_path, const geometry_msgs::msg::TwistStamped & curr_speed)
+{
+  // Rotate in place using max angular velocity / acceleration possible
+  const double sign = angle_to_path > 0.0 ? 1.0 : -1.0;
+  angular_vel = sign * max_angular_vel_;
+
+  // Calculation of angle required for stopping robot with deacceleration in desired angle
+  double deceleration_angle = 0.5 *
+    std::pow(curr_speed.twist.angular.z, 2) / max_angular_accel_ + goal_yaw_tol_;
+
+  if (deceleration_angle > fabs(angle_to_path)) {
+    // sets desired angular velocity
+    angular_vel = 0.0;
+  }
+
+  const double & dt = control_duration_;
+  const double min_feasible_angular_speed = curr_speed.twist.angular.z - max_angular_accel_ * dt;
+  const double max_feasible_angular_speed = curr_speed.twist.angular.z + max_angular_accel_ * dt;
+  angular_vel = std::clamp(angular_vel, min_feasible_angular_speed, max_feasible_angular_speed);
+  RCLCPP_DEBUG(
+    logger_, "speed = %f, speed_cur = %f, angle = %f, calc angle = %f",
+    angular_vel,
+    curr_speed.twist.angular.z,
+    angle_to_path,
+    deceleration_angle);
+}
+
+geometry_msgs::msg::PoseStamped ShimController::getLookAheadPoint(
+  const double & lookahead_dist,
+  const nav_msgs::msg::Path & transformed_plan)
+{
+  // Find the first pose which is at a distance greater than the lookahead distance
+  auto goal_pose_it = std::find_if(
+    transformed_plan.poses.begin(), transformed_plan.poses.end(), [&](const auto & ps) {
+      return hypot(ps.pose.position.x, ps.pose.position.y) >= lookahead_dist;
+    });
+
+  // If the no pose is not far enough, take the last pose
+  if (goal_pose_it == transformed_plan.poses.end()) {
+    goal_pose_it = std::prev(transformed_plan.poses.end());
+  }
+
+  return *goal_pose_it;
+}
+
+void ShimController::setPlan(const nav_msgs::msg::Path & path)
+{
+  global_plan_ = path;
+  if (default_plugin) {default_plugin->setPlan(global_plan_);}
+}
+
+void ShimController::setSpeedLimit(
+  const double & speed_limit,
+  const bool & percentage)
+{
+  if (speed_limit == nav2_costmap_2d::NO_SPEED_LIMIT) {
+    // Restore default value
+    desired_linear_vel_ = max_linear_vel_;
+  } else {
+    if (percentage) {
+      // Speed limit is expressed in % from maximum speed of robot
+      desired_linear_vel_ = max_linear_vel_ * speed_limit / 100.0;
+    } else {
+      // Speed limit is expressed in absolute value
+      desired_linear_vel_ = speed_limit;
+    }
+  }
```

Comment:
Where is this logic used? It looks, like `desired_linear_vel_` is unused at all: we could just call `default_plugin->setSpeedLimit()`, I think